mcontrolcenter: init at 0.4.1

Tommimon opened the pull request:
MControlCenter provides a relatively full-featured control panel for MSI laptops, written in Qt. It allows users to control MSI laptop power plans, fan speed, battery charging limits, keyboard backlight settings, USB power sharing and more.

The application comes with strong desktop integration: shortcuts, taskbar item, etc. etc. -- [screenshots here](https://github.com/dmitry-s93/MControlCenter/blob/main/README.md)

The package comes with a patch that corrects MControlCenter's path to the `modprobe` binary, allowing it to automatically load the `ec_sys` kernel module (included with NixOS by default) with the right settings when launched.

This package has been previous requested by the NixOS community (https://github.com/NixOS/nixpkgs/issues/244881). Another pull request (https://github.com/NixOS/nixpkgs/pull/251816) was submitted for this same package. The pull request has been reviewed but I contacted the author personally and he is not interested in completing it any more. For this reason I'm opening a new pull request incorporating all the changes suggested in the review. 

This NixOS package has been tested on a NixOS Plasma5 and GNOME 45 desktops.

fixes:  #244881

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
## Description of changes

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3847

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-december/1711/21

nim65s commented on pkgs/os-specific/linux/mcontrolcenter/default.nix:
```diff
@@ -0,0 +1,65 @@
+  src = fetchFromGitHub {
+    owner = "dmitry-s93";
+    repo = "MControlCenter";
+    rev = "f04ca9878f0ccfa5ea849a3bd17bcb79a5f52a0a";
```
```suggestion
    rev = finalAttrs.version;
```

nim65s commented:
Result of `nixpkgs-review pr 306216` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mcontrolcenter</li>
  </ul>
</details>

nim65s commented on pkgs/os-specific/linux/mcontrolcenter/default.nix:
```diff
@@ -0,0 +1,65 @@
+  patches = [ ./modprobe.patch ];
+
+  cmakeFlags = [
+    "-DENABLE_TESTING=OFF"
```
Why are you disabling tests ? Does this project even has test, or include CTest, for `ENABLE_TESTING` to be used for anything ?

nim65s commented on pkgs/os-specific/linux/mcontrolcenter/default.nix:
```diff
@@ -0,0 +1,65 @@
+
+  cmakeFlags = [
+    "-DENABLE_TESTING=OFF"
+    "-DENABLE_INSTALL=ON"
```
This flag doesn't exist: https://github.com/search?q=repo%3Admitry-s93%2FMControlCenter%20ENABLE_INSTALL&type=code

nim65s commented on pkgs/os-specific/linux/mcontrolcenter/default.nix:
```diff
@@ -0,0 +1,65 @@
+    hash = "sha256-SV78OVRGzy2zFLT3xqeUtbjlh81Z97PVao18P3h/8dI=";
+  };
+
+  patches = [ ./modprobe.patch ];
```
```suggestion
  postPatch = ''
    substituteInPlace src/helper/helper.cpp \
      --replace-warn "/usr/sbin/modprobe" "${pkgs.kmod}/bin/modprobe
  '';
```

nim65s replied:
also, `kmod` should be added in `buildInputs`

nim65s reviewed (changes requested):
I guess you could move this to `pkgs/by-name`. I think using `libsForQt5.callPackage` is not needed, and you can maybe directly extract packages from `libsForQt5`. Could you try ?

Tommimon commented:
> I guess you could move this to `pkgs/by-name`. I think using `libsForQt5.callPackage` is not needed, and you can maybe directly extract packages from `libsForQt5`. Could you try ?

I applied all the suggested changes and tested the package on `x86_64-linux`, everything should work as intended. Here is a list of what I did:
- move to `pkgs/by-name`
- avoid using `libsForQt5.callPackage` by extracting packages from `libsForQt5`
- getting hash from `finalAttrs.version`
- remove unused cmake flags
- patching with `substituteInPlace`
- add `kmod` to build inputs

@nim65s con you confirm the pool request looks fine to you now?

nim65s reviewed (approved):
This looks good to me, thanks !
I do not know if we have a better solution for this DBus .service file.

nim65s commented:
Discussions on matrix raised that maybe this .service file (and maybe the desktop file btw) should be proposed upstream. Maybe you can contribute a PR there with those files ?

Also, using pkgs.writeText instead of those echo would be a bit cleaner.

LordGrimmauld commented:
Built and tested, launches fine on linux-x86-64 and wayland once the ec-sys kernel module is actually loaded.
However, most options are pretty much locked to something with no way to edit them, with the exception of fan curves. Whether that is due to my extensive customization on tlp, acpi and similar i have not yet fully tested.

Tommimon commented:
> Built and tested, launches fine on linux-x86-64 (msi GF75 thin) once the ec-sys kernel module is actually loaded. However, most options are pretty much locked to something with no way to edit them, with the exception of fan curves. Whether that is due to my extensive customization on tlp, acpi and similar i have not yet fully tested.

It seams to be an [upstream issue](https://github.com/dmitry-s93/MControlCenter/issues/161), someone else reported the same problems for the same model which apparently is not fully supported.

For me the battery options are working correctly, it might be a problem related to `tlp` in your case (as you suggested).

LordGrimmauld commented:
Ah upstream lists my precise model (GF75 thin 10SDR) as fully unsupported currently, soo i guess thats my hardware just not being supported then

LordGrimmauld commented on pkgs/by-name/mc/mcontrolcenter/package.nix:
```diff
@@ -0,0 +1,64 @@
+    install -Dm644  $src/resources/mcontrolcenter.svg $out/share/icons/hicolor/32x32/apps/mcontrolcenter.svg
+    install -Dm644 $src/src/helper/mcontrolcenter-helper.conf $out/share/dbus-1/system.d/mcontrolcenter-helper.conf
+    mkdir -p $out/share/dbus-1/system-services
+    echo "[D-BUS Service]" >> $out/share/dbus-1/system-services/mcontrolcenter.helper.service
+    echo "Name=mcontrolcenter.helper" >> $out/share/dbus-1/system-services/mcontrolcenter.helper.service
+    echo "Exec=$out/libexec/mcontrolcenter-helper" >> $out/share/dbus-1/system-services/mcontrolcenter.helper.service
+    echo "User=root" >> $out/share/dbus-1/system-services/mcontrolcenter.helper.service
```
```suggestion
      install -Dm444 $src/src/helper/mcontrolcenter.helper.service $out/share/dbus-1/system-services/mcontrolcenter.helper.service
      substituteInPlace $out/share/dbus-1/system-services/mcontrolcenter.helper.service --replace "/usr" "$out"
```

nim65s replied:
Nice, thanks ! But `--replace` is deprecated: https://github.com/NixOS/nixpkgs/pull/260776

LordGrimmauld replied:
fair call, use the replacements then. I went copying my own code for which i did not know it was deprecated when writing it.

LordGrimmauld reviewed (changes requested):
Looking good, i would however actually make use of the upstream dbus service file and use substituteInPlace to fix the paths on it.

Also remember to squash your commits eventually, you want two commits:
1) `maintainers: add Tommimon` where you *only* add the entry in the maintainers file
2) `mcontrolcenter: init at 0.4.1` where all the rest is squashed into it
The maintainer addition should come first in the history.

LordGrimmauld commented:
> Discussions on matrix raised that maybe this .service file (and maybe the desktop file btw) should be proposed upstream. Maybe you can contribute a PR there with those files ?
> 
> Also, using pkgs.writeText instead of those echo would be a bit cleaner.

In an effort to learn more about nix, i tried doing just that; turns out getting the $out path for `mcontrolcenter` into `writeText` is either nearly or completely impossible, at which point you'd have to use `substituteInPlace` during `installPhase` anyways. Hence there is no reason why you shouldn't just take [the upstream dbus service config](https://github.com/dmitry-s93/MControlCenter/blob/main/src/helper/mcontrolcenter.helper.service) and fix that using `substituteInPlace` instead.

Tommimon committed:
mantainers: add Tommimon

LordGrimmauld reviewed (approved):
Looking good now :+1:

Tommimon commented:
I added the suggested changes:
- squashed everything into two commits
- using upstream service file
- using `substituteInPlace` to patch the `Exec` path
- using `--replace-fail` in every other occurrence of `substituteInPlace`

nim65s commented on pkgs/by-name/mc/mcontrolcenter/package.nix:
```diff
@@ -0,0 +1,61 @@
+    install -Dm644 $src/resources/mcontrolcenter.svg $out/share/icons/hicolor/32x32/apps/mcontrolcenter.svg
+    install -Dm644 $src/src/helper/mcontrolcenter-helper.conf $out/share/dbus-1/system.d/mcontrolcenter-helper.conf
+    install -Dm644 $src/src/helper/mcontrolcenter.helper.service $out/share/dbus-1/system-services/mcontrolcenter.helper.service
+    substituteInPlace $out/share/dbus-1/system-services/mcontrolcenter.helper.service --replace-fail "/usr" "$out"
```
this looks like a patch to me, so if this could go with the other `substituteInPlace` inside `postPatch`, it would be even better I guess.

Tommimon replied:
> this looks like a patch to me, so if this could go with the other `substituteInPlace` inside `postPatch`, it would be even better I guess.

Moved `substituteInPlace` inside `postPatch`, also updated `installPhase` to use the patched version of `mcontrolcenter.helper.service`

nim65s reviewed (approved):
Good ! I still have another nit suggestion, but feel free to ignore it if you are tired of those changes 😅

LordGrimmauld reviewed (approved):
approving again because weganks bot will discard my approval otherwise

Aleksanaa commented on pkgs/by-name/mc/mcontrolcenter/package.nix:
```diff
@@ -0,0 +1,60 @@
+    homepage = "https://github.com/dmitry-s93/MControlCenter";
+    description = "Tool to change the settings of MSI laptops running Linux";
+    license = licenses.gpl3Plus;
+    platforms = with platforms; linux;
```
```suggestion
    platforms = platforms.linux;
```

Aleksanaa commented on pkgs/by-name/mc/mcontrolcenter/package.nix:
```diff
@@ -0,0 +1,60 @@
+    runHook postInstall
+  '';
+
+  meta = with lib; {
```
Don't use `finalAttrs` with `meta = with lib;`, as `with` doesn't shadow.

LordGrimmauld replied:
yeah thats pretty much the recommended solution too

Tommimon commented:
@Aleksanaa, can I ask you if you approve the current state of this pr? After I applied the last changes.

SuperSandro2000 commented on pkgs/by-name/mc/mcontrolcenter/package.nix:
```diff
@@ -0,0 +1,60 @@
+  pname = "mcontrolcenter";
+  version = "0.4.1";
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "MControlCenter";
+      exec = "mcontrolcenter";
+      icon = "mcontrolcenter";
+      comment = finalAttrs.meta.description;
+      desktopName = "MControlCenter";
+      categories = [ "System" ];
+    })
+  ];
+
+  buildInputs = [
+    libsForQt5.qtbase
+    kmod
+  ];
+
+  nativeBuildInputs = [
+    libsForQt5.wrapQtAppsHook
+    libsForQt5.qttools
+    copyDesktopItems
+    cmake
+  ];
+
+  src = fetchFromGitHub {
+    owner = "dmitry-s93";
+    repo = "MControlCenter";
+    rev = finalAttrs.version;
+    hash = "sha256-SV78OVRGzy2zFLT3xqeUtbjlh81Z97PVao18P3h/8dI=";
+  };
+
+  postPatch = ''
+    substituteInPlace src/helper/helper.cpp --replace-fail "/usr/sbin/modprobe" "${kmod}/bin/modprobe"
+    substituteInPlace src/helper/mcontrolcenter.helper.service --replace-fail "/usr" "$out"
+  '';
```
```suggestion
  src = fetchFromGitHub {
    owner = "dmitry-s93";
    repo = "MControlCenter";
    rev = finalAttrs.version;
    hash = "sha256-SV78OVRGzy2zFLT3xqeUtbjlh81Z97PVao18P3h/8dI=";
  };

  postPatch = ''
    substituteInPlace src/helper/helper.cpp \
      --replace-fail "/usr/sbin/modprobe" "${kmod}/bin/modprobe"
    substituteInPlace src/helper/mcontrolcenter.helper.service \
      --replace-fail "/usr" "$out"
  '';

  desktopItems = [
    (makeDesktopItem {
      name = "MControlCenter";
      exec = "mcontrolcenter";
      icon = "mcontrolcenter";
      comment = finalAttrs.meta.description;
      desktopName = "MControlCenter";
      categories = [ "System" ];
    })
  ];

  nativeBuildInputs = [
    libsForQt5.wrapQtAppsHook
    libsForQt5.qttools
    copyDesktopItems
    cmake
  ];

  buildInputs = [
    libsForQt5.qtbase
    kmod
  ];
```

Tommimon replied:
These changes should be applied in latest commit

Aleksanaa commented on pkgs/by-name/mc/mcontrolcenter/package.nix:
```diff
@@ -0,0 +1,62 @@
+    runHook preInstall
+    install -Dm755 mcontrolcenter $out/bin/mcontrolcenter
+    install -Dm755 helper/mcontrolcenter-helper $out/libexec/mcontrolcenter-helper
+    install -Dm644 $src/resources/mcontrolcenter.svg $out/share/icons/hicolor/32x32/apps/mcontrolcenter.svg
```
```suggestion
    install -Dm644 $src/resources/mcontrolcenter.svg $out/share/icons/hicolor/scalable/apps/mcontrolcenter.svg
```

Aleksanaa replied:
Also why `$src`? Does that file get removed when building?

Tommimon replied:
I replaced `$src` with the relative path, also I replaced `32x32` with `scalable`

Tommimon committed:
mcontrolcenter: init at 0.4.1
Co-authored-by: Nadim Kobeissi <nadim@symbolic.software>
